SpatialGDK/Source/SpatialGDKEditor/Private/SchemaGenerator/SpatialGDKEditorSchemaGenerator.cpp

```diff
@@ -739,21 +738,51 @@ bool RunSchemaCompiler()
+	FString SchemaDescriptorOutput = FPaths::Combine(CompiledSchemaDir, TEXT("schema.descriptor"));
+
+	IPlatformFile& PlatformFile = FPlatformFileManager::Get().GetPlatformFile();
 
-	// The schema_compiler cannot create folders.
-	if (!FPaths::DirectoryExists(SchemaDescriptorDir))
+	const FString& SchemaCompilerBaseArgs = FString::Printf(TEXT("--schema_path=\"%s\" --schema_path=\"%s\" --descriptor_set_out=\"%s\" --load_all_schema_on_schema_path "), *SchemaDir, *CoreSDKSchemaDir, *SchemaDescriptorOutput);
+
+	// If there's already a compiled schema dir, blow it away so we don't have lingering artifacts from previous generation runs.
+	if (FPaths::DirectoryExists(CompiledSchemaDir))
 	{
-		IPlatformFile& PlatformFile = FPlatformFileManager::Get().GetPlatformFile();
-		if (!PlatformFile.CreateDirectoryTree(*SchemaDescriptorDir))
+		if (!PlatformFile.DeleteDirectoryRecursively(*CompiledSchemaDir))
 		{
-			UE_LOG(LogSpatialGDKSchemaGenerator, Error, TEXT("Could not create schema descriptor directory '%s'! Please make sure the parent directory is writeable."), *SchemaDescriptorDir);
-			return false;
+			UE_LOG(LogSpatialGDKSchemaGenerator, Error, TEXT("Could not delete pre-existing compiled schema directory '%s'! Please make sure the directory is writeable."), *CompiledSchemaDir);
+		}
+	}
+
+	// schema_compiler cannot create folders, so we need to set them up beforehand.
+	if (!PlatformFile.CreateDirectoryTree(*CompiledSchemaDir))
+	{
+		UE_LOG(LogSpatialGDKSchemaGenerator, Error, TEXT("Could not create compiled schema directory '%s'! Please make sure the parent directory is writeable."), *CompiledSchemaDir);
+		return false;
+	}
+
+	FString AdditionalSchemaCompilerArgs;
+
+	TArray<FString> Tokens;
+	TArray<FString> Switches;
+	FCommandLine::Parse(FCommandLine::Get(), Tokens, Switches);
+
+	if (const FString* schemaCompileArgsCLSwitchPtr = Switches.FindByPredicate([](const FString& clSwitch) { return clSwitch.StartsWith(FString{ TEXT("AdditionalSchemaCompilerArgs") }); }))
+	{
+		FString switchName;
+		schemaCompileArgsCLSwitchPtr->Split(FString{ TEXT("=") }, &switchName, &AdditionalSchemaCompilerArgs);
+		if (AdditionalSchemaCompilerArgs.Contains(FString{ TEXT("ast_proto_out") }) || AdditionalSchemaCompilerArgs.Contains(FString{ TEXT("ast_json_out") }))
+		{
+			if (!PlatformFile.CreateDirectoryTree(*CompiledSchemaASTDir))
+			{
+				UE_LOG(LogSpatialGDKSchemaGenerator, Error, TEXT("Could not create compiled schema AST directory '%s'! Please make sure the parent directory is writeable."), *CompiledSchemaASTDir);
+				return false;
+			}
 		}
 	}
 
-	FString SchemaCompilerArgs = FString::Printf(TEXT("--schema_path=\"%s\" --schema_path=\"%s\" --descriptor_set_out=\"%s\" --load_all_schema_on_schema_path"), *SchemaDir, *CoreSDKSchemaDir, *SchemaDescriptorOutput);
+	FString SchemaCompilerArgs = FString::Printf(TEXT("%s %s"), *SchemaCompilerBaseArgs, *AdditionalSchemaCompilerArgs);
```

Comment:
Might be worth checking that the additional schema args are valid options, and warning otherwise.

Reply:
Do we also want to parse out all the desired target files / directories and make sure those get created as well?

It's starting to feel like the "passthrough" method is more work than the initial pass on this was (give specific flags which dictate the formats you want and let this function resolve the actual compiler args and output paths). Thoughts?

Reply:
Yeah that's reasonable, passthrough allows us to update the compiler without re-working this code. I think it's fine as is.